AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m

```diff
@@ -37,18 +37,41 @@ - (instancetype)init {
     _expirationTimeColumnIndex = ((NSNumber *)columnIndexes[kMSAppDocumentTableName][kMSExpirationTimeColumnName]).unsignedIntegerValue;
     _downloadTimeColumnIndex = ((NSNumber *)columnIndexes[kMSAppDocumentTableName][kMSDownloadTimeColumnName]).unsignedIntegerValue;
     _operationTimeColumnIndex = ((NSNumber *)columnIndexes[kMSAppDocumentTableName][kMSOperationTimeColumnName]).unsignedIntegerValue;
-    _pendingOperationColumnIndex = ((NSNumber *)columnIndexes[kMSAppDocumentTableName][kMSPendingDownloadColumnName]).unsignedIntegerValue;
+    _pendingOperationColumnIndex = ((NSNumber *)columnIndexes[kMSAppDocumentTableName][kMSPendingOperationColumnName]).unsignedIntegerValue;
   }
   return self;
 }
 
 #pragma mark - Table Management
 
-// TODO work item created to track this implementation
-- (BOOL)createWithPartition:(NSString *)__unused partition
-                   document:(MSDocumentWrapper *)__unused document
-               writeOptions:(MSWriteOptions *)__unused writeOptions {
-  return YES;
+- (BOOL)upsertWithPartition:(NSString *)partition
+            documentWrapper:(MSDocumentWrapper *)documentWrapper
+                  operation:(NSString *_Nullable)operation
+                    options:(MSBaseOptions *)options {
+  NSDate *now = [NSDate date];
+  NSDate *expirationTime = [now dateByAddingTimeInterval:options.deviceTimeToLive];
+  NSString *insertQuery =
+      [NSString stringWithFormat:@"REPLACE INTO '%@' ('%@', '%@', '%@', '%@', '%@', '%@', '%@', '%@') "
+                                 @"VALUES ('%@', '%@', '%@', '%@', '%@', '%@', '%@', '%@')",
+                                 kMSAppDocumentTableName, kMSPartitionColumnName, kMSDocumentIdColumnName, kMSDocumentColumnName,
+                                 kMSETagColumnName, kMSExpirationTimeColumnName, kMSDownloadTimeColumnName, kMSOperationTimeColumnName,
+                                 kMSPendingOperationColumnName, partition, documentWrapper.documentId, documentWrapper.jsonValue,
+                                 documentWrapper.eTag, expirationTime, documentWrapper.lastUpdatedDate, now, operation];
+  NSInteger result = [self.dbStorage executeNonSelectionQuery:insertQuery];
+  if (result != SQLITE_OK) {
+    MSLogError([MSDataStore logTag], @"Unable to update or replace stored document, SQLite error code: %ld", (long)result);
+  }
+  return result == SQLITE_OK;
+}
+
+- (BOOL)deleteWithPartition:(NSString *)partition documentId:(NSString *)documentId {
+  NSString *deleteQuery = [NSString stringWithFormat:@"DELETE FROM '%@' WHERE '%@' = '%@' AND '%@' = '%@'", kMSAppDocumentTableName,
```

Comment:
If it is deleted locally immediately, with no pending operation, how does cosmos db know about the deletion? I guess it must only be marked with the "delete" pending operation for now, and then when synchronizing with cosmos db, deleted locally after it is deleted remotely